### PR TITLE
Add the login class and the test

### PR DESCRIPTION
**Pull request by Robchaves**
A part of the login class is added to the project and the test for the functions in it, this class isn't finished yet.

### CODE REVIEW

File: localMatchApp/headers/Login.h [hunk 1]
```diff
@@ -0,0 +1,44 @@
+/**
+ * \class Login
+ *
+ * \brief Create the connection with the database and the core.
+ *
+ * This class check out the format of the password and check if the
+ * user exists, in the sign-up function.
+ *
+ *
```
Comment (luvalverde) [inline localMatchApp/headers/Login.h]: remove extra blank lines

File: localMatchApp/headers/Login.h [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+  ~Login ( );
+  Login ( const Login& ) = default;
+  Login& operator= ( const Login& ) = default;
+
+  /** This method checks the format of the password when a user wants to sign-up in the
+   * application, the correct format of a password has minimum a digit, a capital letter,
+   * and a small letter.*/
+  bool hasCorrectFormat (const std::string &i_password );
+
+  /** This method is created for add a new user in the application database*/
```
Comment (luvalverde) [inline localMatchApp/headers/Login.h]: This method is created for adding a ...

File: localMatchApp/headers/Login.h [hunk 3]
```diff
@@ -0,0 +1,44 @@
...
+  ConnectionMessage signUp (const std::string &i_user, const std::string &i_password,
+                             const std::string &i_repeatedPassword );
```
Comment (luvalverde) [inline localMatchApp/headers/Login.h:38]: add a blank line after this line of code to separate functions

File: localMatchApp/sources/Login.cpp [hunk 4]
```diff
@@ -0,0 +1,82 @@
+#include "Login.h"
+#include <iostream>
+
+using namespace std;
+
+Login::Login ( ) {
+  // ToDo: Connect to database.
+  // This console out is created for debug process
+  std::cout<<"object created"<<std::endl;
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: please remove all your cout for debugging purposes

File: localMatchApp/sources/Login.cpp [hunk 5]
```diff
@@ -0,0 +1,82 @@
+#include "Login.h"
+#include <iostream>
+
+using namespace std;
+
+Login::Login ( ) {
+  // ToDo: Connect to database.
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: Please change all your ToDo for TODO

File: localMatchApp/sources/Login.cpp [hunk 6]
```diff
@@ -0,0 +1,82 @@
...
+Login::~Login ( ) {
+  // ToDo: Disconnect to database.
+  // This console out is created for debug process
+  std::cout<<"object deleted"<<std::endl;
+}
+
+Login::ConnectionMessage Login::signUp ( const string& i_user, const string& i_password,
+                                         const string& i_repeatedPassword ) {
+  ConnectionMessage signUpConnectionState=ConnectionMessage::goodEntry;
+  // ToDo: Send and recive information to database.
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: TODO: Send and receive information from the database.

File: localMatchApp/sources/Login.cpp [hunk 7]
```diff
@@ -0,0 +1,82 @@
...
+
+  if ( !hasCorrectFormat ( i_password ) || !hasCorrectFormat ( i_repeatedPassword ) ) {
+    signUpConnectionState= ConnectionMessage::wrongPassword;
+  }
+
+  if ( i_password != i_repeatedPassword && (hasCorrectFormat ( i_password ) && hasCorrectFormat ( i_repeatedPassword ))) {
+    signUpConnectionState=ConnectionMessage::notEqualPasswords;
+  }
+
+  if ( i_user==userInDatabase || i_user.length()==0) {
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: it is not necessary to check for the length of the username.
If it is zero, the hasCorrectFormat will return false.

File: localMatchApp/sources/Login.cpp [hunk 8]
```diff
@@ -0,0 +1,82 @@
...
+  ConnectionMessage signUpConnectionState=ConnectionMessage::goodEntry;
+  // ToDo: Send and recive information to database.
+  // This string is just created for debug process
+  string userInDatabase = "Lol2018";
+
+  if ( !hasCorrectFormat ( i_password ) || !hasCorrectFormat ( i_repeatedPassword ) ) {
+    signUpConnectionState= ConnectionMessage::wrongPassword;
+  }
+
+  if ( i_password != i_repeatedPassword && (hasCorrectFormat ( i_password ) && hasCorrectFormat ( i_repeatedPassword ))) {
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: why are you checking again the password format (hasCorrectFormat ) ?

File: localMatchApp/sources/Login.cpp [hunk 9]
```diff
@@ -0,0 +1,82 @@
...
+  // ToDo: Search the user and compare password
+  // ToDo: Query asking for password based on user
+  // These strings are just created for debug process
+  string userInDatabase = "Lol2018";
+  string passInDatabase = "Lol2018";
+
+  if ( i_user != userInDatabase ) {
+    signUpConnectionState= ConnectionMessage::doesntExists;
+  }
+  if ( i_password != passInDatabase && (i_user == userInDatabase) ) {
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: why are you checking again (i_user == userInDatabase) ?
Use IF, ELSE IF statements

File: localMatchApp/sources/Login.cpp [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+      if ( islower ( i ) ) {
+        hasASmallLetter= true;
+      }
+
+      if ( isdigit ( i ) ) {
+        hasANumber= true;
+      }
+    }
+
+  if ( hasANumber && hasACapital&& hasASmallLetter) {
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: This IF statement could be inside the FOR loop, so you don't have to iterate through all the string to know if the format is valid or not

File: localMatchApp/test/logintest.cpp [hunk 11]
```diff
@@ -0,0 +1,83 @@
+#include <gmock/gmock-matchers.h>
+#include <gtest/gtest.h>
+#include <iostream>
+#include <string>
+#include "Login.h"
+using namespace testing;
+
+TEST ( LoginTest, TestPasswordGoodCases ) {
```
Comment (luvalverde) [inline localMatchApp/test/logintest.cpp:10]: provide a test description of what you are trying to verify

File: localMatchApp/test/logintest.cpp [hunk 12]
```diff
@@ -0,0 +1,83 @@
...
+  Login objectOfTest;
+
+  std::string pass1 = "Robert0";
```
Comment (luvalverde) [inline localMatchApp/test/logintest.cpp]: these are not good variable names

File: localMatchApp/test/logintest.cpp [hunk 13]
```diff
@@ -0,0 +1,83 @@
...
+
+  EXPECT_TRUE ( objectOfTest.hasCorrectFormat ( pass1 ) );
+  EXPECT_TRUE ( objectOfTest.hasCorrectFormat ( pass2 ) );
+  EXPECT_TRUE ( objectOfTest.hasCorrectFormat ( pass3 ) );
+  EXPECT_TRUE ( objectOfTest.hasCorrectFormat ( pass4 ) );
+  EXPECT_TRUE ( objectOfTest.hasCorrectFormat ( pass5 ) );
+  EXPECT_TRUE ( objectOfTest.hasCorrectFormat ( pass6 ) );
+}
+
+TEST ( LoginTest, TestPasswordBadCases ) {
```
Comment (luvalverde) [inline localMatchApp/test/logintest.cpp:30]: provide a test description of what you are trying to verify

File: localMatchApp/test/logintest.cpp [hunk 14]
```diff
@@ -0,0 +1,83 @@
...
+
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass1 ) );
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass2 ) );
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass3 ) );
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass4 ) );
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass5 ) );
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass6 ) );
+}
+
+TEST(LoginTest, TestSignUp){
```
Comment (luvalverde) [inline localMatchApp/test/logintest.cpp:49]: provide a test description of what you are trying to verify

File: localMatchApp/test/logintest.cpp [hunk 15]
```diff
@@ -0,0 +1,83 @@
...
+  Login::ConnectionMessage notEqualPasswords= Login::ConnectionMessage::notEqualPasswords;
+
+  std::string emptyUserspace = "";
+  std::string sameUser = "Lol2018";
+  std::string newUser = "newUser";
+  std::string goodPass = "Lol2018";
+  std::string wrongPassGoodFormat = "Il0";
+  std::string BadPass = "lol";
+
+  EXPECT_EQ(objectOfTest.signUp(newUser,goodPass,goodPass),goodEntry);
```
Comment (luvalverde) [inline localMatchApp/test/logintest.cpp]: the left parameter of the EXPECT_EQ must be the EXPECTED value.
The parameter on the right side must be the value OBTAINED.

Your parameters are switched.

File: localMatchApp/test/logintest.cpp [hunk 16]
```diff
@@ -0,0 +1,83 @@
...
+  EXPECT_EQ(objectOfTest.signUp(sameUser,goodPass,goodPass),invalidUser);
+  EXPECT_EQ(objectOfTest.signUp(emptyUserspace,goodPass,goodPass),invalidUser);
+  EXPECT_EQ(objectOfTest.signUp(newUser,BadPass,goodPass),wrongPassword);
+  EXPECT_EQ(objectOfTest.signUp(newUser,goodPass,BadPass),wrongPassword);
+  EXPECT_EQ(objectOfTest.signUp(newUser,goodPass,wrongPassGoodFormat),notEqualPasswords);
+}
+
+TEST(LoginTest, TestEntry){
```
Comment (luvalverde) [inline localMatchApp/test/logintest.cpp:68]: provide a test description of what you are trying to verify

File: localMatchApp/test/logintest.cpp [hunk 17]
```diff
@@ -0,0 +1,83 @@
...
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass3 ) );
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass4 ) );
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass5 ) );
+  EXPECT_FALSE ( objectOfTest.hasCorrectFormat ( wrongPass6 ) );
+}
+
+TEST(LoginTest, TestSignUp){
+  Login objectOfTest;
+
+  Login::ConnectionMessage goodEntry= Login::ConnectionMessage::goodEntry;
```
Comment (luvalverde) [inline localMatchApp/test/logintest.cpp]: it is not necessary to create variables for all of the enum values.
You can use the enum value directly inside the EXPECT_EQ

**Review by luvalverde — CHANGES_REQUESTED** [review]
I added some changes

File: localMatchApp/headers/Login.h (hunk 1 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h]: Done.

File: localMatchApp/headers/Login.h (hunk 2 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h]: Done.

File: localMatchApp/headers/Login.h (hunk 3 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h:38]: Done.

File: localMatchApp/sources/Login.cpp (hunk 4 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

File: localMatchApp/headers/Login.h [hunk 18]
```diff
@@ -0,0 +1,44 @@
...
+ *
+ */
+
+#ifndef LOGIN_H
+#define LOGIN_H
+
+#include <string>
+class Login {
+ public:
+  /// Create a class type enum to communicate the state of the methods "signUp" and "entry".
```
Comment (wfallascr) [inline localMatchApp/headers/Login.h]: Change comment. First part is unnecessary "Create a class type enum to communicate..."

File: localMatchApp/sources/Login.cpp (hunk 5 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

File: localMatchApp/headers/Login.h [hunk 19]
```diff
@@ -0,0 +1,44 @@
...
+  enum class ConnectionMessage {
+    notEqualPasswords,
+    goodEntry,
+    doesntExists,
+    invalidUser,
+    wrongPassword
+  };
+  /// The constructor connects with the database.
```
Comment (wfallascr) [inline localMatchApp/headers/Login.h]: Same as before. "The constructor ..." is unnecessary. Also, the specific comment of "connects with the database" should go inside the constructor

File: localMatchApp/headers/Login.h [hunk 20]
```diff
@@ -0,0 +1,44 @@
...
+  Login ( );
+  /// The destructor disconnects the database.
+  ~Login ( );
+  Login ( const Login& ) = default;
+  Login& operator= ( const Login& ) = default;
+
+  /** This method checks the format of the password when a user wants to sign-up in the
```
Comment (wfallascr) [inline localMatchApp/headers/Login.h]: Same as before "This method  ..."

File: localMatchApp/headers/Login.h (hunk 2 above)
Comment (wfallascr) [reply localMatchApp/headers/Login.h]: "This method is created for..." is irrelevant

**Review by wfallascr — CHANGES_REQUESTED** [review]

File: localMatchApp/headers/Login.h (hunk 2 above)
Comment (luvalverde) [reply localMatchApp/headers/Login.h]: What @wfallascr is trying to say is that you don't have to write in the description that "this is a method" since it is irrelevant or explicit in the code.

File: localMatchApp/test/logintest.cpp (hunk 13 above)
Comment (Robchaves) [reply localMatchApp/test/logintest.cpp:30]: Done.

File: localMatchApp/test/logintest.cpp (hunk 14 above)
Comment (Robchaves) [reply localMatchApp/test/logintest.cpp:49]: Done.

File: localMatchApp/test/logintest.cpp (hunk 15 above)
Comment (Robchaves) [reply localMatchApp/test/logintest.cpp]: Done.

File: localMatchApp/test/logintest.cpp (hunk 16 above)
Comment (Robchaves) [reply localMatchApp/test/logintest.cpp:68]: Done.

File: localMatchApp/test/logintest.cpp (hunk 17 above)
Comment (Robchaves) [reply localMatchApp/test/logintest.cpp]: Done.

File: localMatchApp/headers/Login.h (hunk 18 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h]: Done.

File: localMatchApp/headers/Login.h (hunk 19 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h]: Done.

File: localMatchApp/headers/Login.h (hunk 20 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h]: Done.

**Comment by Robchaves** [comment]
All changes applied, PR back to you.
@luvalverde  @wfallascr

File: localMatchApp/sources/Login.cpp (hunk 8 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: In the test, it was a corner case, the "if" checks the difference of the passwords but both have to be in the correct format.

File: localMatchApp/sources/Login.cpp (hunk 7 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: The "hasCorrectFormat" function just check the format of the password.

File: localMatchApp/sources/Login.cpp (hunk 9 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: I thought that I can't nest if statements.

File: localMatchApp/sources/Login.cpp (hunk 8 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: You are right, change applied.

File: localMatchApp/sources/Login.cpp (hunk 9 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: You are right, change applied.

File: localMatchApp/sources/Login.cpp (hunk 10 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: You are right, change applied.

**Comment by luvalverde** [comment]
@Robchaves is this PR ready for review?
I still do not see several changes being applied, like the IF statements we talked about.

File: localMatchApp/test/logintest.cpp (hunk 12 above)
Comment (Robchaves) [reply localMatchApp/test/logintest.cpp]: Variable names changed

**Comment by Robchaves** [comment]
All changes applied, PR back to you.
@luvalverde @wfallascr

File: localMatchApp/headers/Login.h [hunk 21]
```diff
@@ -0,0 +1,43 @@
+/**
+ * \class Login
+ *
+ * \brief Create the connection with the database and the core.
+ *
+ * This class check out the format of the password and check if the
```
Comment (luvalverde) [inline localMatchApp/headers/Login.h]: Remove all "This class" comments in your classes

File: localMatchApp/sources/Login.cpp [hunk 22]
```diff
@@ -0,0 +1,79 @@
...
+}
+
+Login::~Login ( ) {
+  // TODO: Disconnect to database.
+}
+
+Login::ConnectionMessage Login::signUp ( const string& i_user, const string& i_password,
+                                         const string& i_repeatedPassword ) {
+  ConnectionMessage signUpConnectionState=ConnectionMessage::goodEntry;
+  // TODO: Send and recive information to database.
```
Comment (luvalverde) [inline localMatchApp/sources/Login.cpp]: receive

**Review by luvalverde — CHANGES_REQUESTED** [review]
Minor comments from my part @Robchaves

File: localMatchApp/headers/Login.h (hunk 21 above)
Comment (Robchaves) [reply localMatchApp/headers/Login.h]: Done.

File: localMatchApp/sources/Login.cpp (hunk 22 above)
Comment (Robchaves) [reply localMatchApp/sources/Login.cpp]: Done.

**Comment by Robchaves** [comment]
All changes applied, PR back to you.
@luvalverde @wfallascr

**Review by wfallascr — APPROVED** [review]

**Comment by Robchaves** [comment]
PR back to you.
@wfallascr @luvalverde

File: localMatchApp/test/logintest.cpp (hunk 11 above)
Comment (Robchaves) [reply localMatchApp/test/logintest.cpp:10]: Done.

**Review by luvalverde — APPROVED** [review]
